vpr/src/place/place.cpp

```diff
@@ -1695,6 +1946,82 @@ static void update_td_delta_costs(const PlaceDelayModel* delay_model,
+
+    for (ClusterPinId clb_pin : blocks_affected.affected_pins) {
+        //It is possible that some connections may not have changed delay.(e.g.
+        //For instance, if using a dx/dy delay model, this could occur if a sink
+        //moved to a new position with the same dx/dy from it's driver.
+        //
+        //To minimize work during the incremental STA update we do not invalidate
+        //such unchanged connections.
+
+        ClusterNetId net = clb_nlist.pin_net(clb_pin);
+        int ipin = clb_nlist.pin_net_index(clb_pin);
+
+        if (proposed_connection_delay[net][ipin] != connection_delay[net][ipin]) {
+            //Delay has changed. Must invalidate this sink pin.
+            sink_pins_affected.push_back(clb_pin);
+        }
+    }
+}
+
+/**
+ * @brief Check if the setup slack has gotten better or worse due to block swap.
+ *
+ * Get all the modified slack values via the PlacerSetupSlacks class, and compare
+ * then with the original values at these connections. Sort them and compare them
+ * one by one, and return the difference of the first different pair.
+ *
+ * If the new slack value is larger(better), than return a negative value so that
+ * the move will be accepted. If the new slack value is smaller(worse), return a
+ * positive value so that the move will be rejected.
+ *
+ * If no slack values have changed, then return an arbitrary positive number. A
+ * move resulting in no change in the slack values should probably be unnecessary.
+ *
+ * The sorting is need to prevent in the unlikely circumstances that a bad slack
+ * value suddenly got very good due to the block move, while a good slack value
+ * got very bad, perhaps even worse than the original worse slack value.
+ */
+static float analyze_setup_slack_cost(const PlacerSetupSlacks* setup_slacks) {
+    const auto& cluster_ctx = g_vpr_ctx.clustering();
+    const auto& clb_nlist = cluster_ctx.clb_nlist;
+
+    //Find the original/proposed setup slacks of pins with modified values
+    std::vector<float> original_setup_slacks, proposed_setup_slacks;
+
+    auto clb_pins_modified = setup_slacks->pins_with_modified_setup_slack();
+    for (ClusterPinId clb_pin : clb_pins_modified) {
+        ClusterNetId net_id = clb_nlist.pin_net(clb_pin);
+        size_t ipin = clb_nlist.pin_net_index(clb_pin);
+
+        original_setup_slacks.push_back(connection_setup_slack[net_id][ipin]);
+        proposed_setup_slacks.push_back(setup_slacks->setup_slack(net_id, ipin));
+    }
+
+    //Sort in ascending order, from the worse slack value to the best
+    std::sort(original_setup_slacks.begin(), original_setup_slacks.end());
+    std::sort(proposed_setup_slacks.begin(), proposed_setup_slacks.end());
+
+    //Check the first pair of slack values that are different
+    //If found, return their difference
+    for (size_t idiff = 0; idiff < original_setup_slacks.size(); ++idiff) {
+        float slack_diff = original_setup_slacks[idiff] != proposed_setup_slacks[idiff];
```

Comment:
Should this be:

```
float slack_diff = original_setup_slacks[idiff] - proposed_setup_slacks[idiff];
```

(subtract instead of not equals)

Reply:
Thanks for the catch!